src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelModTidyFunction.java

```diff
@@ -0,0 +1,119 @@
+import com.google.devtools.build.lib.vfs.RootedPath;
+import com.google.devtools.build.skyframe.SkyFunction;
+import com.google.devtools.build.skyframe.SkyFunctionException;
+import com.google.devtools.build.skyframe.SkyKey;
+import com.google.devtools.build.skyframe.SkyValue;
+import com.google.devtools.build.skyframe.SkyframeLookupResult;
+import javax.annotation.Nullable;
+import net.starlark.java.eval.EvalException;
+
+/**
+ * Computes all information required for the {@code bazel mod tidy} command. The evaluation of all
+ * module extensions used by the root module is triggered to, as a side effect, emit any {@link
+ * RootModuleFileFixupEvent}s.
+ */
+public class BazelModTidyFunction implements SkyFunction {
+
+  @Override
+  @Nullable
+  public SkyValue compute(SkyKey skyKey, Environment env)
+      throws InterruptedException, SkyFunctionException {
+    BazelDepGraphValue depGraphValue = (BazelDepGraphValue) env.getValue(BazelDepGraphValue.KEY);
+    if (depGraphValue == null) {
+      return null;
+    }
+    RepositoryMappingValue bazelToolsRepoMapping =
+        (RepositoryMappingValue)
+            env.getValue(RepositoryMappingValue.key(RepositoryName.BAZEL_TOOLS));
+    if (bazelToolsRepoMapping == null) {
+      return null;
+    }
+    Label buildozerLabel;
+    try {
+      buildozerLabel =
+          Label.parseWithRepoContext(
+              "@buildozer_binary//:buildozer.exe",
+              Label.RepoContext.of(
+                  RepositoryName.BAZEL_TOOLS, bazelToolsRepoMapping.getRepositoryMapping()));
+    } catch (LabelSyntaxException e) {
+      throw new IllegalStateException(e);
+    }
+    RootedPath buildozer;
+    try {
+      buildozer = RepositoryFunction.getRootedPathFromLabel(buildozerLabel, env);
+    } catch (NeedsSkyframeRestartException e) {
+      return null;
+    } catch (EvalException e) {
+      throw new IllegalStateException(e);
+    }
+
+    ImmutableSet<SkyKey> extensionsUsedByRootModule =
+        depGraphValue.getExtensionUsagesTable().columnMap().get(ModuleKey.ROOT).keySet().stream()
+            .map(SingleExtensionEvalValue::key)
+            .collect(toImmutableSet());
+    SkyframeLookupResult result = env.getValuesAndExceptions(extensionsUsedByRootModule);
+    if (env.valuesMissing()) {
+      return null;
+    }
+    for (SkyKey extension : extensionsUsedByRootModule) {
+      try {
+        result.getOrThrow(extension, ExternalDepsException.class);
+      } catch (ExternalDepsException e) {
+        if (e.getDetailedExitCode().getFailureDetail() == null
+            || !e.getDetailedExitCode()
+                .getFailureDetail()
+                .getExternalDeps()
+                .getCode()
+                .equals(FailureDetails.ExternalDeps.Code.INVALID_EXTENSION_IMPORT)) {
+          throw new BazelModTidyFunctionException(e, SkyFunctionException.Transience.PERSISTENT);
+        }
+        // This is an error bazel mod tidy can fix, so don't fail.
```

Comment:
As far as I understand this, we can't recover *in a SkyFunction*, but here we perform a top-level evaluation (implicitly in keep going mode: https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/skyframe/SkyframeExecutor.java;l=2262;drc=0e1dd2afc8dfe69033abbc7852bb23bd0a9237f6). The individual values will still be marked as failed, but we should be able to extract the individual exceptions in this way. There is a test case covering this case, but if I misunderstood how this works it might of course not catch every issue due to the inherent non-determinism of Skyframe evaluation order.

Reply:
ah! I think the 'implicitly in keep-going mode' bit is crucial. But it still surprises me a bit that the whole evaluation doesn't end up in error (even though we can go through all the failed nodes). Anyhow, this is definitely not blocking the whole PR, especially as you do have a test case

Reply:
> But it still surprises me a bit that the whole evaluation doesn't end up in error (even though we can go through all the failed nodes).

You are right, that is indeed surprising. Maybe this is because the nodes we actually requested did not fail themselves and we are in "keep going" mode? It would be nice to get confirmation on this.